File: java-pi-pantry/src/test/java/com/langrsoft/pi/pantry/PantryTest.java
package com.langrsoft.pi.pantry;

import static com.langrsoft.util.LambdaUtil.map;
import static java.util.Arrays.asList;
import static java.util.Collections.emptyList;
import static java.util.stream.Collectors.toList;
import static org.hamcrest.Matchers.*;
import static org.junit.Assert.*;
import java.time.*;
import java.util.*;
import org.junit.Test;
import com.langrsoft.util.TestClock;

/**
 * // Tests for Pantry class
 *
 * author: I'll never tell
 */
public class PantryTest {
   static final LocalDate NOW = LocalDate.now();

   private Pantry pantry = new Pantry();

   @Test
   public void contains1() {
      try {
         Pantry pantry = new Pantry();
         Item i1 = new Item("cheerios");
         pantry.purchase(i1);
         assertThat(pantry.contains("cheerios"), equalTo(true));
      } catch (IllegalArgumentException e) {
         fail("purchase failed:" + e.getMessage());
      }
   }

   /**
    * contains2
    */
   @Test
   public void contains2() {
      try {
         Pantry pantry = new Pantry();
         Item cheerios = new Item("cheerios");
         cheerios.setDescription("cereal");
         pantry.purchase(cheerios);
         assertThat(pantry.contains("sugar"), is(equalTo(false)));
      } catch (IllegalArgumentException e) {
         fail("purchase failed:" + e.getMessage());
      }
   }

   /**
    * count2
    *
    * count of items by name
    */
   @Test
   public void count() {
      try {
         Pantry pantry = new Pantry();
         assertThat(pantry.count("sugar"), is(equalTo(0)));

         pantry.purchase(new Item("sugar"));

         pantry.purchase(new Item("cheerios"));

         pantry.purchase(new Item("cheerios"));
         assertThat("pantry count not correct for cheerioes", pantry.count("cheerios"), is(equalTo(2)));

      } catch (IllegalArgumentException e) {
         fail("purchase failed:" + e.getMessage());
      }

   }

   /**
    * count
    *
    * count of all items
    */
   @Test
   public void count2() {
      Pantry pantry = new Pantry();
      pantry.purchase(new Item("sugar"));
      pantry.purchase(new Item("sugar"));
      pantry.purchase(new Item("cheerios"));
      pantry.purchase(new Item("cheerios"));
      assertThat("pantry count not 4", pantry.count(), is(equalTo(4)));
   }

   /**
    * dt
    *
    * ensure purchase date set to today
    */
   @Test
   public void get0() {
      Pantry pantry = new Pantry();
      Item sugar = new Item("sugar");
      sugar.setDescription("refined sweetener");
      pantry.purchase(sugar);
      Item retrieved = pantry.getItemNamed("sugar");
      assertThat("item should not be null", retrieved, is(not(nullValue()))); // null?
      assertThat(retrieved.getDescription(), is(equalTo("refined sweetener")));
   }

   /**
    * dt
    *
    * ensure purchase date set to today
    * uses the java clock class as a basis for the fake
    */
   @Test
   public void dt() {
      LocalDateTime todayAtNoon = LocalDate.now().atTime(12, 0, 0);
      Instant todayNoonUTC = todayAtNoon.toInstant(ZoneOffset.UTC);
      Clock clock = Clock.fixed(todayNoonUTC, ZoneOffset.UTC);
      Pantry pantry = new Pantry();
      // inject fake clock
      pantry.setClock(clock);
      Item eggs = new Item("eggs");
      eggs.setDescription("oval ovum");
      pantry.purchase(eggs);
      Item retrieved = pantry.getItemNamed("eggs");
      assertThat(retrieved, is(not(nullValue())));
      assertThat(retrieved.getPurchaseDate(), is(equalTo(LocalDate.now())));
   }

   @Test
   public void get3() {
      Pantry pantry = new Pantry();
      assertThat(pantry.getItemNamed("did not purchase"), is(nullValue()));
   }

   @Test
   public void purchaseNull() {
      try {
         Pantry pantry = new Pantry();
         pantry.purchase(null);
         fail("should have thrown IllegalArgumentException");
      } catch (IllegalArgumentException e) {
         // ok
      }
   }

   @Test
   public void get2() {
      Pantry pantry = new Pantry();
      Item smallCoffee = new Item("coffee");
      smallCoffee.setDescription("small");
      Item largeCoffee = new Item("coffee");
      largeCoffee.setDescription("large");
      pantry.purchase(smallCoffee);
      pantry.purchase(largeCoffee);
      List<Item> coffees = pantry.getItemsNamed("coffee");
      assertThat(coffees.stream().map(Item::getDescription).collect(toList()),
         equalTo(asList("small", "large")));
      List<Item> items = pantry.getItemsNamed("pizza");
      assertThat(items, is(emptyList()));
   }

   @Test
   public void expieringToday() {
      Pantry pantry = new Pantry();
      LocalDate d = LocalDate.now();
      pantry.setClock(TestClock.fixedTo(d));
      Item item1 = new Item("milk");
      item1.setCategory("milk");
      item1.setExpirationDate(d);
      pantry.purchase(item1);
      // add another item
      Item item2 = new Item("blood orange juice 24oz");
      item2.setExpirationDate(d.plusDays(1));
      item2.setCategory("orange juice");
      pantry.purchase(item2);
      // retrieve
      List<Item> items = pantry.getItemsExpiringToday();
      List<String> actualItemNames = new ArrayList<>();
      for (Item i: items) {
         actualItemNames.add(i.getName());
      }
      System.out.println("actual item names:" + actualItemNames);
      List<String> expectedItemNames = new ArrayList<>();
      expectedItemNames.add("milk");
      assertThat(actualItemNames, equalTo(expectedItemNames));
   }


   @Test
   public void allExpiredItemsIncludesItemsExpiringTodayOrEarlier() {
      pantry.setClock(TestClock.fixedTo(NOW));
      pantry.purchase(new ItemBuilder("milk").withExpirationDate(NOW).create());
      pantry.purchase(new ItemBuilder("juice").withExpirationDate(NOW.plusDays(1)).create());
      pantry.purchase(new ItemBuilder("curdled milk").withExpirationDate(NOW.minusDays(1)).create());

      List<Item> items = pantry.getAllExpiredItems();

      assertThat(map(items, Item::getName),
         containsInAnyOrder("milk", "curdled milk"));
   }

   @Test
   public void tooMany() {
      Pantry pantry = new Pantry();
      for (int i = 0; i < 512; i++) {
         Item item = new Item();
         pantry.purchase(item);
      }
      try {
         Item item = new Item();
         pantry.purchase(item);
         fail("expected exception");
      } catch (RuntimeException e) {
         // ok
      }
   }
}

File: csharp-pi-pantry/csharp-pi-pantry-test/DomainTests/PantryTest.cs
using System;
using System.Collections.Generic;
using NUnit.Framework;
using Pipantry.Domain;
using System.Linq;
using Moq;
using Pipantry.Util;

namespace Test.Pipantry.Domain
{
    [TestFixture]
    /**
     * // Tests for Pantry class
     *
     * author: I'll never tell
     */
    public class PantryTest
    {
        [Test]
        public void Contains1()
        {
            try
            {
                var p = new Pantry();
                var i1 = new Item("cheerios");
                p.Purchase(i1);
                Assert.That(p.Contains("cheerios"), Is.EqualTo(true));
            }
            catch (ArgumentException e)
            {
                Assert.Fail("purchase Failed:" + e.Message);
            }
        }

        /**
         * contains2
         */
        [Test]
        public void Contains2()
        {
            try
            {
                var pantry = new Pantry();
                var cheerios = new Item("cheerios");
                cheerios.Description = "cereal";
                pantry.Purchase(cheerios);
                Assert.That(pantry.Contains("sugar"), Is.EqualTo(false));
            }
            catch (ArgumentException e)
            {
                Assert.Fail("purchase Failed:" + e.Message);
            }
        }

        /**
         * count2
         *
         * count of items by name
         */
        [Test]
        public void Count()
        {
            try
            {
                var p = new Pantry();
                //Assert.That(p.Count("sugar"), Is.Zero);

                p.Purchase(new Item("sugar"));

                p.Purchase(new Item("cheerios"));

                p.Purchase(new Item("cheerios"));
                Assert.That(p.Count("cheerios"), Is.EqualTo(2), "pantry count not correct for cheerios");

            }
            catch (ArgumentException e)
            {
                Assert.Fail("purchase Failed:" + e.Message);
            }
        }

        /**
         * count
         *
         * count of all items
         */
        [Test]
        public void Count2()
        {
            var pantry = new Pantry();
            pantry.Purchase(new Item("sugar"));
            pantry.Purchase(new Item("sugar"));
            pantry.Purchase(new Item("cheerios"));
            pantry.Purchase(new Item("cheerios"));
            Assert.That(pantry.Count(), Is.EqualTo(4), "pantry count not 4");
        }

        /**
         * dt
         *
         * ensure purchase date set to today
         */
        [Test]
        public void Get0()
        {
            var pantry = new Pantry();
            var sugar = new Item("sugar");
            sugar.Description = "refined sweetener";
            pantry.Purchase(sugar);
            var retrieved = pantry.ItemNamed("sugar");
            Assert.That(retrieved, Is.Not.Null, "item should not be null"); // null?
            Assert.That(retrieved.Description, Is.EqualTo("refined sweetener"));
        }

        /**
         * dt
         *
         * ensure purchase date set to today
         * uses the the clock class as a basis for the fake
         */
        [Test]
        public void Dt()
        {
            DateTime now = DateTime.Now;
            DateTime todayAtNoon = new DateTime(now.Year, now.Month, now.Day, now.Hour, now.Minute, now.Second);
            var pantry = new Pantry();
            var mock = new Mock<IDateTime>();
            mock.Setup(p => p.Now).Returns(todayAtNoon);

            // inject fake clock
            pantry.DateTimeProvider = mock.Object;

            var eggs = new Item("eggs");
            eggs.Description = "oval ovum";
            pantry.Purchase(eggs);
            var retrieved = pantry.ItemNamed("eggs");
            Assert.That(retrieved, Is.Not.Null);
            Assert.That(retrieved.PurchaseDate, Is.EqualTo(todayAtNoon));
        }

        [Test]
        public void Get3()
        {
            var pant = new Pantry();
            Assert.That(pant.ItemNamed("did not purchase"), Is.Null);
        }

        [Test]
        public void PurchaseNull()
        {
            try
            {
                var pantry = new Pantry();
                pantry.Purchase(null);
                Assert.Fail("should have thrown ArgumentException");
            }
            catch (ArgumentException e)
            {
                // ok
            }
        }

        [Test]
        public void Get2()
        {
            var pantry = new Pantry();
            var smallCoffee = new Item("coffee");
            smallCoffee.Description = "small";
            var largeCoffee = new Item("coffee");
            largeCoffee.Description = "large";
            pantry.Purchase(smallCoffee);
            pantry.Purchase(largeCoffee);
            var coffees = pantry.ItemsNamed("coffee");
            Assert.That(coffees.Select(item => item.Description),
                    Is.EquivalentTo(new string[] { "small", "large" }));
            var items = pantry.ItemsNamed("pizza");
            Assert.That(items, Is.Empty);
        }

        [Test]
        public void ExpieringToday()
        {
            var pantry = new Pantry();
            var now = DateTime.Now;

            var mock = new Mock<IDateTime>();
            mock.Setup(p => p.Now).Returns(now);

            // inject fake clock
            pantry.DateTimeProvider = mock.Object;

            var item1 = new Item("milk");
            item1.Category = "milk";
            item1.ExpirationDate = now;
            pantry.Purchase(item1);
            // add another item
            var item2 = new Item("blood orange juice 24oz");
            item2.ExpirationDate = now.AddDays(1);
            item2.Category = "orange juice";
            pantry.Purchase(item2);
            // retrieve
            var items = pantry.ItemsExpiringToday();
            var actualItemNames = new List<string>();
            foreach (var i in items)
            {
                actualItemNames.Add(i.Name);
            }
            Console.WriteLine("actual item names:" + actualItemNames);
            var expectedItemNames = new List<string>();
            expectedItemNames.Add("milk");
            Assert.That(actualItemNames,
                    Is.EqualTo(expectedItemNames));
        }

        [Test]
        public void AllExpiredItems()
        {
            var pantry = new Pantry();
            var d = DateTime.Now;
            //pantry.setClock(TestClock.fixedTo(d));
            var item1 = new Item("milk");
            item1.Category = "milk";
            item1.ExpirationDate = d;
            pantry.Purchase(item1);
            // add another item
            var item2 = new Item("blood orange juice 24oz");
            item2.ExpirationDate = d.AddDays(1);
            item2.Category = "orange juice";
            pantry.Purchase(item2);
            string name = "curdled milk";
            var item3 = new Item(name);
            item3.ExpirationDate = d.AddDays(-1);
            item1.Category = "milk";
            pantry.Purchase(item3);
            // retrieve
            var items = pantry.getAllExpiredItems();
            var actualItemNames = new List<string>();
            foreach (var i in items)
            {
                actualItemNames.Add(i.Name);
            }
            var expectedItemNames = new List<string>();
            expectedItemNames.Add(name);
            expectedItemNames.Add("milk");
            Assert.That(actualItemNames, Is.EquivalentTo(new string[] { "milk", "curdled milk" }));
        }

        [Test]
        public void TooMany()
        {
            var pantry = new Pantry();
            for (var i = 0; i < 512; i++)
            {
                var item = new Item("");
                pantry.Purchase(item);
            }
            try
            {
                var item = new Item("");
                pantry.Purchase(item);
                Assert.Fail("expected exception");
            }
            catch (Exception e)
            {
                // ok
            }
        }
    }
}
